### PR TITLE
Implement dataclass for EdgeObject

### PR DESCRIPTION
**Pull request by fantix**
This would allow using the `EdgeObject` as a dataclass, in particular:

* `dataclasses.is_dataclass(obj)` will return `True`
* `dataclasses.asdict(obj)` will return a proper `dict`
* No impact to performance

Then we could do something like this with FastAPI as an example:

```python
import fastapi
import edgedb
import uuid
from dataclasses import dataclass

app = fastapi.FastAPI()
client = edgedb.create_client()

@dataclass
class Role:
    id: uuid.UUID
    name: str

    @classmethod
    def __get_validators__(cls):
        return []

@dataclass
class User:
    id: uuid.UUID
    name: str
    role: Role

    @classmethod
    def __get_validators__(cls):
        return []

@app.get("/", response_model=User)
def read_root():
    return client.query("select User { name, role: { name } }")[0]
```

Schema:

```
module default {
    type User {
        property name -> str;
        link role -> Role;
    }

    type Role {
        required property name -> str;
    }
}
```

### CODE REVIEW

**Comment by i0bs** [comment]
Have not tested, but for this part:

```py
return client.query("select User { name, role: { name } }")[0]
```

Will the first index return us the `Role` Python dataclass, or the EdgeQL schema of it?

**Comment by fantix** [comment]
> Have not tested, but for this part:
> 
> ```python
> return client.query("select User { name, role: { name } }")[0]
> ```
> 
> Will the first index return us the `Role` Python dataclass, or the EdgeQL schema of it?

Oh that's the first object in the result set as we are using `query()` instead of `query_single()`.

**Comment by fantix** [comment]
Furthermore, we will be able to separate this into 2 separate files:

`user_role.edgeql`:

```
select User { name, role: { name } }
```

`user_role_edgeql.py`:

```python
import edgedb
import typing
import uuid
from dataclasses import dataclass

@dataclass
class Role:
    id: uuid.UUID
    name: str

    @classmethod
    def __get_validators__(cls):
        return []

@dataclass
class User:
    id: uuid.UUID
    name: str
    role: Role

    @classmethod
    def __get_validators__(cls):
        return []

def query(client: edgedb.Client) -> typing.Sequence[User]:
    return client.query("select User { name, role: { name } }")
```

So that we can generate `user_role_edgeql.py` from `user_role.edgeql` with a console command.

**Comment by i0bs** [comment]
> Oh that's the first object in the result set as we are using `query()` instead of `query_single()`.
> 
> p.s.: Ideally we could do something like:
> 
> ```python
> @app.get("/", response_model=typing.Iterable[User])
> def read_root():
>     return client.query("select User { name, role: { name } }")
> ```
> 
> But we're blocked by [tiangolo/fastapi#3913](https://github.com/tiangolo/fastapi/pull/3913) - so using a regular `list` is the current alternative.

That would be preferred, or you could subsequently use `typing.Mapping` to make a relationship between the dataclass and then client-defined typings. Would that also work?

**Comment by fantix** [comment]
> subsequently use `typing.Mapping` to make a relationship between the dataclass and then client-defined typings. Would that also work?

An example please? Sorry I'm not sure I'm 100% following.

**Comment by i0bs** [comment]
> An example please? Sorry I'm not sure I'm 100% following.

`typing.Mapping` works [like a container,](https://docs.python.org/3/glossary.html#term-mapping) you essentially would have an object that returns a value type based on the key type:

```py
class CustomMapping(typing.Generic[KV, KT]):
    def __getitem__(self, key: KT) -> VT:
        ...
```

This would only be useful if you have custom-defined typings in the client library. For example, [you could use type variables](https://docs.python.org/3/library/typing.html#typing.TypeVar) to represent types from EdgeQL.

File: edgedb/datatypes/record_desc.c [hunk 1]
```diff
@@ -30,6 +30,7 @@ record_desc_dealloc(EdgeRecordDescObject *o)
     PyObject_GC_UnTrack(o);
     Py_CLEAR(o->index);
     Py_CLEAR(o->names);
+    Py_CLEAR(o->get_dataclass_fields);
```
Comment (1st1) [inline edgedb/datatypes/record_desc.c:33]: Do we need to visit it in `record_desc_traverse`?

File: edgedb/datatypes/record_desc.c [hunk 2]
```diff
@@ -177,12 +178,23 @@ record_desc_dir(EdgeRecordDescObject *o, PyObject *args)
 }
 
 
+static PyObject *
+record_set_dataclass_fields_func(EdgeRecordDescObject *o, PyObject *arg)
+{
+    o->get_dataclass_fields = arg;
```
Comment (1st1) [inline edgedb/datatypes/record_desc.c:184]: Py_CLEAR it first before the assignment?

Comment (1st1) [inline edgedb/datatypes/record_desc.c:184]:
```suggestion
    o->get_dataclass_fields_func = arg;
```

File: edgedb/datatypes/record_desc.c (hunk 1 above)
Comment (1st1) [reply edgedb/datatypes/record_desc.c:33]: OK, this is a function.. the name is confusing. I suggest renaming to `get_dataclass_fields_func`

File: edgedb/datatypes/object.c [hunk 3]
```diff
@@ -189,6 +189,10 @@ object_hash(EdgeObject *o)
 static PyObject *
 object_getattr(EdgeObject *o, PyObject *name)
 {
+    // Used in `dataclasses.as_dict()`
+    if (PyUnicode_CompareWithASCIIString(name, "__dataclass_fields__") == 0) {
+        return EdgeRecordDesc_GetDataclassFields((PyObject *)o->desc);
+    }
```
Comment (1st1) [inline edgedb/datatypes/object.c:195]: I think this should be moved into the `L_NOT_FOUND` case. Pointless to call this on every other attribute lookup.

File: edgedb/datatypes/object.c [hunk 4]
```diff
@@ -365,6 +369,15 @@ EdgeObject_InitType(void)
         return NULL;
     }
 
+    // Pass the `dataclasses.is_dataclass(obj)` check, the dict is always empty
+    PyObject *default_fields = PyDict_New();
+    if (default_fields == NULL) {
+        return NULL;
+    }
+    PyDict_SetItemString(
+        EdgeObject_Type.tp_dict, "__dataclass_fields__", default_fields
```
Comment (1st1) [inline edgedb/datatypes/object.c:378]: Why is this needed if we have a getter?

Comment (fantix) [reply edgedb/datatypes/object.c:378]: In order to pass the `dataclasses.is_dataclass(obj)` check, which checks `hasattr(type(obj), "__dataclass_fields__")`

Comment (1st1) [reply edgedb/datatypes/object.c:384]: > which checks hasattr(type(obj), "__dataclass_fields__")

Can you add a comment about that?

**Review by 1st1 — APPROVED** [review]

File: edgedb/datatypes/object.c (hunk 4 above)
Comment (fantix) [reply edgedb/datatypes/object.c:385]: Sure - the comment is more complete now
